src/main/sensors/gyroanalyse.c

```diff
@@ -255,59 +265,101 @@ static FAST_CODE_NOINLINE void gyroDataAnalyseUpdate(gyroAnalyseState_t *state,
+            }
+            // lower Max value to catch first peak close to max
+            dataMax = 0.8f * dataMax;
+            dataAvg = dataAvg / FFT_BIN_COUNT;
+            // automatically set peak detection threshold at half difference between peak and average
+            dynFiltThreshold = 0.5f * (dataMax / dataAvg);
             // iterate over fft data and calculate weighted indices
             for (int i = 1 + fftBinOffset; i < FFT_BIN_COUNT; i++) {
                 const float data = state->fftData[i];
                 const float prevData = state->fftData[i - 1];
-
-                if (fftIncreasing || data > prevData * FFT_MIN_BIN_RISE) {
-                    float cubedData = data * data * data;
-
-                    // add previous bin before first rise
-                    if (!fftIncreasing) {
-                        cubedData += prevData * prevData * prevData;
-
-                        fftIncreasing = true;
+                // disregard fft bins after first peak
+                if (!fftPeakFinished) {
+                    // include bins around the first bin that exceeds 80% max bin height and increased compared to previous bin 
+                    if (fftIncreasing || ((data > prevData * dynFiltThreshold) && (data > dataMax))) {
+                        float cubedData = data * data * data;
+                        // add previous bin
+                        if (!fftIncreasing) {
+                            cubedData += prevData * prevData * prevData;
+                            fftIncreasing = true;
+                        }
+                        // peak over when incoming bin falls below average
+                        if (data < dataAvg) {
+                            fftPeakFinished = true;
+                        }
+                        fftSum += cubedData;
+                        // calculate weighted index starting at 1, not 0
+                        fftWeightedSum += cubedData * (i + 1);
                     }
-
-                    fftSum += cubedData;
-                    // calculate weighted index starting at 1, not 0
-                    fftWeightedSum += cubedData * (i + 1);
                 }
             }
 
             // get weighted center of relevant frequency range (this way we have a better resolution than 31.25Hz)
             // if no peak, go to highest point to minimise delay
-            float centerFreq = dynNotchMaxCentreHz;
+            float centerFreq = dynamicNotchMaxCenterHz;
             float fftMeanIndex = 0;
-
             if (fftSum > 0) {
                 // idx was shifted by 1 to start at 1, not 0
                 fftMeanIndex = (fftWeightedSum / fftSum) - 1;
                 // the index points at the center frequency of each bin so index 0 is actually 16.125Hz
-                centerFreq = constrain(fftMeanIndex * fftResolution, DYN_NOTCH_MIN_CENTRE_HZ, dynNotchMaxCentreHz);
+                centerFreq = fftMeanIndex * fftResolution;
+            } else {
```

Comment:
According to code above it can never be 0

Reply:
This part of the code i have not touched.

Reply:
You have added an `else` branch

Reply:
What should I do to fix this? Should I add back the constraint?

Reply:
OK the else branch, now I know better what your concern is .

Reply:
I'm just saying that `else` branch seems unreachable because `fftSum` is never 0

Reply:
Ah, OK, I put it in as a safety, only in case the previous peak detection code simply found no peak at all.  This is possible if the bin heights smoothly get less and less.  Then no successive FFT bin gets tall enough to start getting added into fftSum.  In that case, it holds the previous centre value.  I think it is necessary.  Some quads just don't have a peak.